push_config.go

```diff
@@ -0,0 +1,176 @@
+// If this attribute has been supplied, the identities passed to 'IdentityFilters' must be a subset of the identities passed to
+// 'Identities'; passing in identities that are not provided in the 'Identities' field will be ignored.
+type PushConfig struct {
+	Identities      map[string][]EventType `msgpack:"identities" json:"identities"`
+	IdentityFilters map[string]string      `msgpack:"filters"    json:"filters"`
+	Params          url.Values             `msgpack:"parameters" json:"parameters"`
+
+	// parsedIdentityFilters holds the parsed `IdentityFilters` to avoid re-parsing the configured filters on each push
+	// event that is using the same config.
+	parsedIdentityFilters map[string]*Filter
+}
+
+// NewPushFilter returns a new PushFilter. NewPushFilter is now aliased to NewPushConfig. This was done for backwards
+// compatibility as a result of the re-naming of PushFilter to PushConfig.
+//
+// Deprecated: use the constructor with the new name, NewPushConfig, instead
+func NewPushFilter() *PushFilter {
+	fmt.Println("DEPRECATED: elemental.NewPushFilter is deprecated, use elemental.NewPushConfig instead")
+
+	return &PushFilter{
+		Identities:            map[string][]EventType{},
+		IdentityFilters:       map[string]string{},
+		parsedIdentityFilters: map[string]*Filter{},
+	}
+}
+
+// NewPushConfig returns a new PushConfig.
+func NewPushConfig() *PushConfig {
+
+	return &PushConfig{
+		Identities:            map[string][]EventType{},
+		IdentityFilters:       map[string]string{},
+		parsedIdentityFilters: map[string]*Filter{},
+	}
+}
+
+// SetParameter sets the values of the parameter with the given key.
+func (f *PushConfig) SetParameter(key string, values ...string) {
+
+	if f.Params == nil {
+		f.Params = url.Values{}
+	}
+
+	f.Params[key] = values
+}
+
+// Parameters returns a copy of all the parameters.
+func (f *PushConfig) Parameters() url.Values {
+
+	if f.Params == nil {
+		return nil
+	}
+
+	out := url.Values{}
+	for k, v := range f.Params {
+		out[k] = v
+	}
+
+	return out
+}
+
+// FilterIdentity adds the given identity for the given eventTypes in the PushConfig.
+func (f *PushConfig) FilterIdentity(identityName string, eventTypes ...EventType) {
+
+	f.Identities[identityName] = eventTypes
+}
+
+// ParseIdentityFilters does something...
+//
+// TODO:
+//  - add a proper comment explaining what this API does
+//  - add unit tests
+func (f *PushConfig) ParseIdentityFilters() error {
+
+	for identity, unparsedFilter := range f.IdentityFilters {
+		if _, found := f.Identities[identity]; !found {
+			return fmt.Errorf("elemental: cannot declare an identity filter on %q as that was not declared in 'Identities'", identity)
+		}
+
+		filter, err := NewFilterParser(unparsedFilter).Parse()
+		if err != nil {
+			return fmt.Errorf("elemental: unable to parse filter %q: %s", unparsedFilter, err)
+		}
+
+		f.parsedIdentityFilters[identity] = filter
+	}
+
+	return nil
+}
```

Comment:
@primalmotion 👀

Reply:
Notice that I am considering it an error to be defining a "fine-grained" identity filter on an identity that you did not declare in the `Identities` attribute.

I think this makes sense because if you are interested in only say, **IdentityA**, then you would declare this as you have done so before. However, now as an additional feature, if you wish to further restrict pushes only to those identities which also have some property (e.g. `name="antoine"`), then you would explicitly declare that too.

**Success example:**

- Identities: 

`A: [get, create]`
`B: [get, create]`


- IdentityFilters:

`A: "someProperty exists"`

This `PushConfig` should (_eventually_) result in receiving only those things that:

- Are of identity types A or B
- If they are of type identity A, then they must have the property "someProperty"


**Error example:**

- Identities: 

`A: [get, create]`
`B: [get, create]`


- IdentityFilters:

`C: "someProperty exists"`

This is an error because you did not declare the identity "C" in your PushConfig `identities`. It is open to interpretation so the safest thing to do is error out.


Your thoughts?

@t00f @primalmotion

Reply:
bump

Reply:
I think I am good with this approach. It makes sense to consider the second case as an error if you said you're only interested by identities A and B.